add classList methods to element

fravan opened the pull request:
Hello there,

I was looking for handling `classList` and stuff but the API was not there, so I added it.

Here is how to use it in gleam code:
```gleam
element.remove_classes(el, ["to-be-deleted", "to-be-removed"])
element.add_class(el, "new-class")
element.add_classes(el, ["and-two", "and-three"])
element.toggle_class(el, "toggled")
element.toggle_class_when(el, "toggled-with-predicate", element.inner_text(el) == "I should toggle as well")
element.replace_class(el, "new-class", "new-class-replaced")
```

Open to any changes, I don't know how _true_ to conventions I am (List vs Array for instance?).

If needed, here is the MDN page: https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

Cheers!

CrowdHailer commented:
Thanks for this.

Have you tested these? Does the Gleam List type work with the spread operator?

For plinth I am preferring Array over List in the API. Plinth should have the minimum amount of opinions and Array is the underlying thing that the functions ask for. 
To that extent I think I would like to create a dom_token_list.gleam module and have all these functions work on that.
i.e. `element.class_list` should return an object to work with. Eventually someone will ask for every function that is available.
Finally. Can you remove the helpers.
there should be only an `add` function that takes an array, because that the API from the DOM/
there should be only one `toggle` which takes a force parameter as second argument. (Also I don't think force and predicate have the same connotations. https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle)

fravan commented:
Thanks for the quick review!

I did test it and it works, because `List` as an iterator symbol. But I will change to `Array` just to keep it consistent with everything else.

Also, do you want to me to look for adding this `element.class_list` and start of a `dom_token_list.gleam` module ? I can give it a look if you want. This would avoid "spoiling" the element module with some of those methods.

I did completely messed up my understanding of the 2nd param of `toggle` indeed. It is not a predicate at all.

_I'm hanging around the Gleam discord server if needed (same name, cat pic)_

CrowdHailer commented on src/plinth/browser/dom_token_list.gleam:
```diff
@@ -0,0 +1,24 @@
+pub fn toggle(
+  dom_token_list: DomTokenList,
+  token: String,
+  force force: Option(Bool),
```
options of bool seems awkward. do Some(False) and Nil behave differently at all. if not juse a bool will do,

fravan replied:
I can't use False. In JS there is a difference between `undefined` and `false` as per the doc:

> If included, turns the toggle into a one way-only operation. If set to false, then token will only be removed, but not added. If set to true, then token will only be added, but not removed.

I could add a custom type to make it clearer ?
```gleam
pub type ToggleOption {
  AddOnly
  RemoveOnly
  Toggle
}
```

Or split into `toggle(DomTokenList, String)` and `toggle_force(DomTokenList, String, Bool)`

What do you think?

CrowdHailer replied:
As there is a difference I think Option(Bool) is actually representative of the API and is good as it is.
Could you add documentation to the Gleam functions. I pretty much just copy paste what it says on mdn.

CrowdHailer commented on src/plinth/browser/element.gleam:
```diff
@@ -170,3 +171,6 @@ pub fn get_checked(element: Element) -> Bool
 @external(javascript, "../../element_ffi.mjs", "contains")
 pub fn contains(element: Element, other: Element) -> Bool
+
+@external(javascript, "../../element_ffi.mjs", "getClassList")
```
I think most of the other api's in plinth are without the get. both in Gleam and the JS helper

CrowdHailer commented:
Thanks, this is much better. I do swing by the discord sometime but not today. there are several people who can give good guidance there.

fravan commented:
I've added comments, and links to MDN doc for more info.

I also updated the API as 2 of those methods return a `Bool` (we learn something every day!)

CrowdHailer commented:
Thank you very much
